README.md
# Polymaker

[![Fork](https://img.shields.io/github/forks/yusriltakeuchi/polymaker?style=social)](https://github.com/yusriltakeuchi/polymaker/fork)&nbsp; [![Star](https://img.shields.io/github/stars/yusriltakeuchi/polymaker?style=social)](https://github.com/yusriltakeuchi/polymaker/star)&nbsp; [![Watches](https://img.shields.io/github/watchers/yusriltakeuchi/polymaker?style=social)](https://github.com/yusriltakeuchi/polymaker/)&nbsp; [![Get the library](https://img.shields.io/badge/Get%20library-pub-blue)](https://pub.dev/packages/polymaker)&nbsp; [![Example](https://img.shields.io/badge/Example-Ex-success)](https://pub.dev/packages/polymaker#-example-tab-)

Polymaker is a flutter package to make it easier to map polygon locations in Google Maps, because the appearance is realtime when in Edit Mode, so we know the exact position without guessing through the backend system.
Polymaker can be run with only one line of code, and returns the location's List value.

### Tools Feature:

- Get Current Location
- Entering Edit Mode
- Closing Edit Mode
- Done Editing
- Undo Editing to Previous Location
- Realtime polygon to view result
- Custom Marker as Pointing Number
- Point Distance
- Tracking Mode LINEAR & PLANAR
- Custom Map Type

<p>
<img src="https://i.imgur.com/pjJdUvw.jpg" height="480px">
<img src="https://i.imgur.com/nfyJked.jpg" height="480px">
</p>
  
# Billing

You must enable some API in google cloud to use this features
- Google Maps for Android/iOS

## Setup

#### ANDROID

Adding Permission in **android/app/src/main/AndroidManifest.xml**

```xml
<uses-permission  android:name="android.permission.INTERNET"/>
<uses-permission  android:name="android.permission.ACCESS_COARSE_LOCATION"/>
<uses-permission  android:name="android.permission.ACCESS_FINE_LOCATION" />
```

Insert Google Maps Key in **android/app/src/main/AndroidManifest.xml**

```xml
<application
	android:name="io.flutter.app.FlutterApplication"
	android:label="example"
	android:icon="@mipmap/ic_launcher">
	.....
	<meta-data  android:name="com.google.android.geo.API_KEY"
		android:value="<YOUR API KEY>"/>
	<meta-data  android:name="com.google.android.gms.version"
		android:value="@integer/google_play_services_version" />
</application>
```

Setup your SDK Version to the latest in **android/app/build.gradle**

    compileSdkVersion 29
    ....
    defaultConfig {
      ....
      targetSdkVersion 29
      ....
    }

#### IOS

Specify your API key in the application delegate `ios/Runner/AppDelegate.m`:

```objectivec
#include "AppDelegate.h"
#include "GeneratedPluginRegistrant.h"
#import "GoogleMaps/GoogleMaps.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application
    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  [GMSServices provideAPIKey:@"YOUR KEY HERE"];
  [GeneratedPluginRegistrant registerWithRegistry:self];
  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}
@end
```

Or in your swift code, specify your API key in the application delegate `ios/Runner/AppDelegate.swift`:

```swift
import UIKit
import Flutter
import GoogleMaps

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?
  ) -> Bool {
    GMSServices.provideAPIKey("YOUR KEY HERE")
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}
```
Opt-in to the embedded views preview by adding a boolean property to the app's `Info.plist` file
with the key `io.flutter.embedded_views_preview` and the value `YES`; you need also to define `NSLocationWhenInUseUsageDescription`

```
  <key>NSLocationWhenInUseUsageDescription</key>
  <string>This app needs your location to test the location feature of the Google Maps location picker plugin.</string>
  <key>io.flutter.embedded_views_preview</key>
  <true/>
```

## How To Use

  Using polymaker is very easy and only needs to use one line of code

```dart
import  'package:polymaker/polymaker.dart'  as polymaker;

// Open polymaker and get return List<LatLng>
var result =  await polymaker.getLocation(context);
```

## Custom Property in getLocation()

|Property|Description |Data Type
|--|--|--|
|**toolColor** |Property to customize tool color |Color
|**polygonColor** |Property to customize polygon color |Color
|**iconLocation** |Property to customize location icon |IconData
|**iconEditMode** |Property to customize edit mode icon |IconData
|**iconCloseEdit** |Property to customize close tool icon |IconData
|**iconDoneEdit** |Property to customize done icon |IconData
|**iconUndoEdit** |Property to customize undo icon |IconData
|**iconGPSPoint** |Property to use GPS data as tracking point |IconData
|**autoEditMode** |Automatic enable edit mode |Boolean
|**pointDistance** |Enable / Disable Point Distance |Boolean
|**trackingMode** |Choose Tracking mode between polygon or polyline |TrackingMode
|**targetCameraPosition** |Set initial camera locations |LatLng
|**enableDragMarker** |Enabling drag to marker position |Boolean

CHANGELOG.md
## [1.0.3] - 2022-05-10

- Updating dependency to latest
- Fix example project deprecated of Androdi Embedding v2

## [1.0.2] - 2021-03-09

- Fixing pubspec score using dartfmt formatting

## [1.0.1] - 2021-03-09

- Updating Changelog and version

## [1.0.0-nullsafety.1] - 2021-03-09

- Migrating package to null safety
- Upgrading another packages
- Change RaisedButton in example to ElevatedButton
- Change location to geolocator
- Removing FadeAnimation class
- Fixing some bugs

## [0.0.5] - 2021-02-22

- Updating dependency to the latest versions
- Updating example project sdk version to 29
- Added new feature Draggable Marker

## [0.0.4] - 2020-05-18

- getLocation return type is now Future<List<LatLng>>
- Added Tracking Mode Options [TrackingMode.LINEAR and TrackingMode.PLANAR]
- Added initial camera locations
- Added Map Type Control
- Added functionality to use GPS data as tracking points

## [0.0.3+1] - 2020-05-18

- Fixing some warning

## [0.0.3] - 2020-05-18

- Adding Point Distance
- Adding AutoEditMode option
- Adding PointDistance option

## [0.0.2+2] - 2020-05-15

- Fixing Readme Error

## [0.0.2+1] - 2020-05-15

- Cleaning & Formatting Some Code
- Updating Readme

## [0.0.2] - 2020-05-15

- Adding Custom Marker as Pointing Number
- Adding support SDK Version 1.17.1

## [0.0.1] - 2020-05-14

- First release package


test/polymaker_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:polymaker/polymaker.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
    // expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/polymaker.dart
library polymaker;

import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:polymaker/core/models/trackingmode.dart';
import 'package:polymaker/ui/screens/map_screen.dart';

class PolyMaker {
  ///BuildContext property
  BuildContext context;

  ///Property to customize tool color
  final Color? toolColor;

  ///Property to customize polygon color
  final Color? polygonColor;

  ///Property to customize location icon
  final IconData? iconLocation;

  ///Property to customize edit mode icon
  final IconData? iconEditMode;

  ///Property to customize close tool icon
  final IconData? iconCloseEdit;

  ///Property to customize done icon
  final IconData? iconDoneEdit;

  ///Property to cusstomize undo icon
  final IconData? iconUndoEdit;

  final IconData? iconGPSPoint;

  ///Property to auto edit mode when maps open
  final bool? autoEditMode;

  ///Property to enable and disable point distance
  final bool? pointDistance;

  ///Property to respond to user-defined camera pos
  final LatLng? targetCameraPosition;

  ///Property to choose tracking mode, you can choose PLANAR or LINEAR
  final TrackingMode? trackingMode;

  ///Property to enable draggable marker
  final bool? enableDragMarker;

  PolyMaker(
      {required this.context,
      this.toolColor,
      this.polygonColor,
      this.iconLocation,
      this.iconEditMode,
      this.iconCloseEdit,
      this.iconDoneEdit,
      this.iconUndoEdit,
      this.iconGPSPoint,
      this.autoEditMode,
      this.pointDistance,
      this.targetCameraPosition,
      this.trackingMode,
      this.enableDragMarker});

  ///Function to open location maker and get result locations
  Future<List<LatLng>?> getLocation() async {
    final result = await Navigator.of(context).push(MaterialPageRoute(
        builder: (context) => MapScreen(
              toolColor: toolColor,
              polygonColor: polygonColor,
              iconLocation: iconLocation,
              iconEditMode: iconEditMode,
              iconCloseEdit: iconCloseEdit,
              iconDoneEdit: iconDoneEdit,
              iconUndoEdit: iconUndoEdit,
              iconGPSPoint: iconGPSPoint,
              autoEditMode: autoEditMode,
              pointDistance: pointDistance,
              trackingMode: trackingMode,
              targetCameraPosition: targetCameraPosition,
              enableDragMarker: enableDragMarker,
            )));
    return result;
  }
}

///Function to open location maker
Future<List<LatLng>?> getLocation(BuildContext context,
    {Color? toolColor,
    Color? polygonColor,
    IconData? iconLocation,
    IconData? iconEditMode,
    IconData? iconCloseEdit,
    IconData? iconDoneEdit,
    IconData? iconUndoEdit,
    IconData? iconGPSPoint,
    bool? autoEditMode,
    bool? pointDistance,
    LatLng? targetCameraPosition,
    TrackingMode? trackingMode,
    bool? enableDragMarker}) async {
  return await PolyMaker(
          context: context,
          toolColor: toolColor != null ? toolColor : Colors.black87,
          polygonColor: polygonColor != null ? polygonColor : Colors.red,
          iconLocation: iconLocation != null ? iconLocation : Icons.my_location,
          iconEditMode:
              iconEditMode != null ? iconEditMode : Icons.edit_location,
          iconCloseEdit: iconCloseEdit != null ? iconCloseEdit : Icons.close,
          iconDoneEdit: iconDoneEdit != null ? iconDoneEdit : Icons.check,
          iconUndoEdit: iconUndoEdit != null ? iconUndoEdit : Icons.undo,
          iconGPSPoint:
              iconGPSPoint != null ? iconGPSPoint : Icons.add_location,
          autoEditMode: autoEditMode != null ? autoEditMode : false,
          pointDistance: pointDistance != null ? pointDistance : true,
          targetCameraPosition:
              targetCameraPosition != null ? targetCameraPosition : null,
          trackingMode:
              trackingMode != null ? trackingMode : TrackingMode.PLANAR,
          enableDragMarker: enableDragMarker != null ? enableDragMarker : false)
      .getLocation();
}


lib/ui/screens/map_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:polymaker/core/models/trackingmode.dart';
import 'package:polymaker/core/viewmodels/map_provider.dart';
import 'package:provider/provider.dart';

class MapScreen extends StatefulWidget {
  ///Property to customize tool color
  final Color? toolColor;

  ///Property to customize polygon color
  final Color? polygonColor;

  ///Property to customize location icon
  final IconData? iconLocation;

  ///Property to customize edit mode icon
  final IconData? iconEditMode;

  ///Property to customize close tool icon
  final IconData? iconCloseEdit;

  ///Property to customize done icon
  final IconData? iconDoneEdit;

  ///Property to cusstomize undo icon
  final IconData? iconUndoEdit;

  final IconData? iconGPSPoint;

  ///Property to auto edit mode when maps open
  final bool? autoEditMode;

  ///Property to enable and disable point distance
  final bool? pointDistance;

  ///Property to choose tracking mode, you can choose PLANAR or LINEAR
  final TrackingMode? trackingMode;

  ///Property to enable draggable marker
  final bool? enableDragMarker;

  ///If this is null it means user is opting to use GPS tracking
  final LatLng? targetCameraPosition;

  MapScreen(
      {this.toolColor,
      this.polygonColor,
      this.iconLocation,
      this.iconEditMode,
      this.iconCloseEdit,
      this.iconDoneEdit,
      this.iconUndoEdit,
      this.iconGPSPoint,
      this.autoEditMode,
      this.pointDistance,
      this.trackingMode,
      this.targetCameraPosition,
      this.enableDragMarker});

  @override
  _MapScreenState createState() => _MapScreenState();
}

class _MapScreenState extends State<MapScreen> {
  String distance = "awal";
  bool isSatellite = false;
  @override
  Widget build(BuildContext context) {
    //To modify status bar
    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle(
        statusBarColor: Colors.transparent,
        statusBarIconBrightness: Brightness.dark));

    return Scaffold(
      body: MultiProvider(
        providers: [
          ChangeNotifierProvider(
            create: (context) => MapProvider(),
          )
        ],
        child: Consumer<MapProvider>(
          builder: (contex, mapProv, _) {
            //Get first location
            if (mapProv.cameraPosition == null &&
                mapProv.onInitCamera == false) {
              if (widget.targetCameraPosition != null) {
                mapProv.initCamera(widget.autoEditMode!, widget.pointDistance,
                    targetCameraPosition: widget.targetCameraPosition,
                    dragMarker: widget.enableDragMarker);
              } else {
                mapProv.initCamera(widget.autoEditMode!, widget.pointDistance,
                    dragMarker: widget.enableDragMarker);
              }
              mapProv.setPolygonColor(widget.polygonColor);
              return Center(
                child: CircularProgressIndicator(),
              );
            }

            return Center(
              child: Stack(
                children: <Widget>[
                  Positioned(top: -300, child: mapDistance()),
                  Positioned(top: -300, child: mapIcon()),
                  mapProv.cameraPosition != null
                      ? Container(
                          width: MediaQuery.of(context).size.width,
                          height: MediaQuery.of(context).size.height,
                          child: GoogleMap(
                            myLocationButtonEnabled: false,
                            myLocationEnabled: true,
                            compassEnabled: false,
                            tiltGesturesEnabled: false,
                            markers: mapProv.markers,
                            mapType: isSatellite
                                ? MapType.satellite
                                : MapType.normal,
                            initialCameraPosition: mapProv.cameraPosition!,
                            onMapCreated: mapProv.onMapCreated,
                            mapToolbarEnabled: false,
                            onTap: (loc) => mapProv.onTapMap(loc,
                                mode: widget.trackingMode),
                            polygons: mapProv.polygons,
                            polylines: mapProv.polylines,
                          ),
                        )
                      : Center(
                          child: CircularProgressIndicator(),
                        ),
                  _toolsList(),
                ],
              ),
            );
          },
        ),
      ),
    );
  }

  ///Widget for custom marker
  Widget mapIcon() {
    return Consumer<MapProvider>(
      builder: (context, mapProv, _) {
        return RepaintBoundary(
          key: mapProv.markerKey,
          child: Container(
            width: 32,
            height: 32,
            decoration: BoxDecoration(
                color: widget.polygonColor, shape: BoxShape.circle),
            child: Center(
              child: Text(
                (mapProv.tempLocation.length + 1).toString(),
                style: TextStyle(
                    color: Colors.white,
                    fontWeight: FontWeight.bold,
                    fontSize: 15),
              ),
            ),
          ),
        );
      },
    );
  }

  Widget mapDistance() {
    return Consumer<MapProvider>(
      builder: (context, mapProv, _) {
        return RepaintBoundary(
          key: mapProv.distanceKey,
          child: Container(
            width: mapProv.distance.length > 6
                ? (mapProv.distance.length >= 9 ? 100 : 80)
                : 64,
            height: 32,
            decoration: BoxDecoration(
                color: widget.toolColor,
                borderRadius: BorderRadius.circular(10)),
            child: Center(
              child: Text(mapProv.distance,
                  style: TextStyle(
                      fontSize: 16,
                      fontWeight: FontWeight.bold,
                      color: Colors.white)),
            ),
          ),
        );
      },
    );
  }

  ///Widget tools list
  Widget _toolsList() {
    return Builder(
      builder: (context) {
        return Consumer<MapProvider>(
          builder: (context, mapProv, _) {
            return SafeArea(
              child: Stack(
                children: <Widget>[
                  Align(
                    alignment: Alignment.topRight,
                    child: Padding(
                        padding: const EdgeInsets.only(top: 30, right: 20),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.end,
                          children: <Widget>[
                            mapProv.isEditMode == true
                                ? InkWell(
                                    onTap: () => mapProv.undoLocation(),
                                    child: Container(
                                      width: 40,
                                      height: 40,
                                      decoration: BoxDecoration(
                                          color: isSatellite
                                              ? Colors.white
                                              : widget.toolColor,
                                          borderRadius:
                                              BorderRadius.circular(50)),
                                      child: Icon(
                                        widget.iconUndoEdit,
                                        color: isSatellite
                                            ? Colors.black87
                                            : Colors.white,
                                      ),
                                    ),
                                  )
                                : SizedBox(),
                            SizedBox(
                                width: mapProv.isEditMode == true ? 10 : 0),
                            mapProv.isEditMode == true
                                ? InkWell(
                                    onTap: () => mapProv.saveTracking(context),
                                    child: Container(
                                      width: 40,
                                      height: 40,
                                      decoration: BoxDecoration(
                                          color: isSatellite
                                              ? Colors.white
                                              : widget.toolColor,
                                          borderRadius:
                                              BorderRadius.circular(50)),
                                      child: Icon(
                                        widget.iconDoneEdit,
                                        color: isSatellite
                                            ? Colors.black87
                                            : Colors.white,
                                      ),
                                    ),
                                  )
                                : SizedBox(),
                            SizedBox(
                                width: mapProv.isEditMode == true ? 10 : 0),
                            InkWell(
                              onTap: () => mapProv.changeEditMode(),
                              child: Container(
                                width: 40,
                                height: 40,
                                decoration: BoxDecoration(
                                    color: isSatellite
                                        ? Colors.white
                                        : widget.toolColor,
                                    borderRadius: BorderRadius.circular(50)),
                                child: Icon(
                                  mapProv.isEditMode == false
                                      ? widget.iconEditMode
                                      : widget.iconCloseEdit,
                                  color: isSatellite
                                      ? Colors.black87
                                      : Colors.white,
                                ),
                              ),
                            ),
                            SizedBox(width: 10),
                            InkWell(
                              onTap: () => mapProv.changeCameraPosition(
                                  mapProv.sourceLocation!),
                              child: Container(
                                width: 40,
                                height: 40,
                                decoration: BoxDecoration(
                                    color: isSatellite
                                        ? Colors.white
                                        : widget.toolColor,
                                    borderRadius: BorderRadius.circular(50)),
                                child: Icon(
                                  widget.iconLocation,
                                  color: isSatellite
                                      ? Colors.black87
                                      : Colors.white,
                                ),
                              ),
                            ),
                            SizedBox(width: 10),
                            InkWell(
                              onTap: () {
                                if (isSatellite) {
                                  isSatellite = false;
                                } else
                                  isSatellite = true;
                                setState(() {});
                              },
                              child: Container(
                                width: 40,
                                height: 40,
                                decoration: BoxDecoration(
                                    color: isSatellite
                                        ? Colors.white
                                        : widget.toolColor,
                                    borderRadius: BorderRadius.circular(50)),
                                child: Icon(
                                  isSatellite ? Icons.map : Icons.satellite,
                                  color: isSatellite
                                      ? Colors.black87
                                      : Colors.white,
                                ),
                              ),
                            ),
                          ],
                        )),
                  ),
                  widget.targetCameraPosition == null && mapProv.isEditMode
                      ? Align(
                          alignment: Alignment.bottomLeft,
                          child: Padding(
                            padding:
                                const EdgeInsets.only(bottom: 30, left: 20),
                            child: InkWell(
                              onTap: () {
                                mapProv.addGpsLocation(
                                    mode: widget.trackingMode);
                              },
                              child: Container(
                                width: 40,
                                height: 40,
                                decoration: BoxDecoration(
                                    color: isSatellite
                                        ? Colors.white
                                        : widget.toolColor,
                                    borderRadius: BorderRadius.circular(50)),
                                child: Icon(
                                  widget.iconGPSPoint,
                                  color: isSatellite
                                      ? Colors.black87
                                      : Colors.white,
                                ),
                              ),
                            ),
                          ),
                        )
                      : SizedBox(height: 0, width: 0),
                ],
              ),
            );
          },
        );
      },
    );
  }
}


lib/ui/Animation/FadeAnimation.dart
import 'package:flutter/material.dart';
import 'package:simple_animations/simple_animations.dart';

class FadeAnimation extends StatelessWidget {
  ///Delay for animations
  final double delay;

  ///Animation child
  final Widget child;

  ///Initialize constructor
  FadeAnimation({this.delay, this.child});

  @override
  Widget build(BuildContext context) {
    final tween = MultiTrackTween(
      [
        Track("opacity")
            .add(Duration(milliseconds: 500), Tween(begin: 0.0, end: 1.0)),
        Track("translateY").add(
            Duration(milliseconds: 500), Tween(begin: -30.0, end: 0.0),
            curve: Curves.linearToEaseOut)
      ],
    );

    return ControlledAnimation(
      delay: Duration(milliseconds: (500 * delay).round()),
      duration: tween.duration,
      tween: tween,
      child: child,
      builderWithChild: (context, child, animation) => Opacity(
        opacity: animation["opacity"],
        child: Transform.translate(
            offset: Offset(0, animation["translateY"]), child: child),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:polymaker/core/models/trackingmode.dart';
import 'package:polymaker/polymaker.dart' as polymaker;

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'PolyMaker Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomeScreen(),
    );
  }
}

class HomeScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("PolyMaker Demo"),
      ),
      body: HomeBody(),
    );
  }
}

class HomeBody extends StatefulWidget {
  @override
  _HomeBodyState createState() => _HomeBodyState();
}

class _HomeBodyState extends State<HomeBody> {
  List<LatLng>? locationList;
  void getLocation() async {
    var result = await polymaker.getLocation(
      context,
      trackingMode: TrackingMode.PLANAR,
      enableDragMarker: true,
    );
    if (result != null) {
      setState(() {
        locationList = result;
      });
    }
  }

  @override
  void initState() {
    super.initState();
    locationList = [];
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: MediaQuery.of(context).size.width,
      height: MediaQuery.of(context).size.height,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          Text(
            "Location Result: \n" +
                (locationList != null
                    ? locationList!
                        .map((val) => "[${val.latitude}, ${val.longitude}]\n")
                        .toString()
                    : ""),
            textAlign: TextAlign.center,
            style: TextStyle(fontSize: 18),
          ),
          SizedBox(height: 10),
          Container(
            height: 45,
            child: ElevatedButton(
              style: ElevatedButton.styleFrom(
                primary: Colors.blue,
                shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(10)),
              ),
              onPressed: () => getLocation(),
              child: Text(
                "Get Polygon Location",
                style: TextStyle(fontSize: 16, color: Colors.white),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
